src/infrasys/utils/time_utils.py

```diff
@@ -0,0 +1,174 @@
+        Python object representing the time duration as a delta.
+
+    Raises
+    ------
+    ValueError
+        If fractional milliseconds are provided (e.g, P0DT30.532S)
+        If the string does not follow the ISO 8601 format.
+
+    See Also
+    --------
+    to_iso_8601: Reverse operation of this function
+
+    Examples
+    --------
+    A simple example for a delta of 10 months.
+
+    >>> delta_str = "P1M"
+    >>> result = from_iso_8601(delta_str)
+    >>> print(result)
+    relativedelta(months=1)
+
+    For a delta of 1 hour
+
+    >>> delta_str = "P0DT1H"
+    >>> result = from_iso_8601(delta_str)
+    >>> print(result)
+    timedelta(hours=1)
+    """
+    for name, regex in REGEX_DURATIONS.items():
+        if match := re.match(regex, duration):
+            if name == "milliseconds":
+                value_float = float(match.group(1))
+                if (value_float * 1_000) % 1 != 0.0:
+                    msg = "Fractional milliseconds are not supported. "
+                    msg += "Provide seconds with a integer number of milliseconds"
+                    raise ValueError(msg)
+                value = value_float * 1_000
+            else:
+                value = int(match.group(1))
+            return DURATION_TO_TYPE[name](**{name: value})
+    else:
+        msg = f"No match found for {duration=}. "
+        msg += "Check `REGEX_DURATIONS` to validate that the format is covered."
+        raise ValueError(msg)
+
+
+def to_iso_8601(duration: timedelta | relativedelta) -> str:
+    """Convert a timedelta or relativedelta object to ISO 8601 duration string.
+
+    Parameters
+    ----------
+    duration: timedelta | relativedelta
+        Python object representing a timedelta
+
+    Returns
+    -------
+    str
+        String representation of the duration using the ISO 8601.
+
+    Raises
+    ------
+    TypeError
+        If the object provided is not either `timedelta` or `relativedelta`.
+
+    ValueError
+        If fractional milliseconds are provided (e.g, P0DT30.532S)
+
+    See Also
+    --------
+    from_iso_8601: Reverse operation of this function
+
+    Examples
+    --------
+    A simple example for a delta of 10 months.
+
+    >>> delta = timedelta(hours=1)
+    >>> result = to_iso_8601(delta)
+    >>> print(result)
+    "P0DT1H"
+
+    For a delta of 1 year
+
+    >>> delta = relativedelta(years=1)
+    >>> result = to_iso_8601(delta)
+    >>> print(result)
+    "P1Y"
+    """
+    if not isinstance(duration, (timedelta, relativedelta)):
+        msg = "Input must be a timedelta or relativedelta object."
+        raise TypeError(msg)
+
+    if isinstance(duration, relativedelta):
+        years = duration.years or 0
+        months = duration.months or 0
+        days = duration.days or 0
+        seconds = duration.hours * 3600 + duration.minutes * 60 + duration.seconds
+        microseconds = duration.microseconds
+    else:  # timedelta
+        years = months = 0
```

Comment:
The Julia side is encoding the equivalent of `P0DT{duration.total_seconds()}S` for regular periods. Are we gaining anything by preserving this? In Julia if you create a time series with `SingleTimeSeries(..., resolution = Dates.Hour(1)` and then call `get_resolution(get_time_series(...))`, you will get a Dates.Millisecond value.

Reply:
You mean the `microseconds`? The only real reason to keep it if if someone has  1ms resolution. The way that they will add it could be either `relativedelta(seconds=0.0010)` or `relativedelta(microseconds=1000)`. If that happens we can still interpret that. At the end, both will be converted to seconds.

Reply:
No, I am referring to whether we want to preserve seconds vs minutes vs hours vs days. I suppose this is more important in Julia because `Dates.Hour` is a different type than `Dates.Minute`. We could make `get_resolution` return Hour if that is what the user passes in. Today we always return Millisecond.

I still think you can avoid the logic for days/hours/minutes and always encode in seconds when the type is time delta.

Reply:
No.